tests/framework/matrix.py

```diff
@@ -0,0 +1,152 @@
+    def disk(self, disk):
+        """Setter for disk artifact."""
+        self._context['disk'] = disk
+
+    @property
+    def microvm(self):
+        """Return the microvm artifact."""
+        return self._context.get('microvm', None)
+
+    @microvm.setter
+    def microvm(self, microvm):
+        """Setter for kernel artifact."""
+        self._context['microvm'] = microvm
+
+    @property
+    def snapshot(self):
+        """Return the snapshot artifact."""
+        return self._context.get('snapshot', None)
+
+    @snapshot.setter
+    def snapshot(self, snapshot):
+        """Setter for snapshot artifact."""
+        self._context['snapshot'] = snapshot
+
+    @property
+    def custom(self):
+        """Return the custom context."""
+        return self._context.get('custom', None)
+
+    @custom.setter
+    def custom(self, custom):
+        """Setter for custom context."""
+        self._context['custom'] = custom
+
+
+class TestMatrix:
+    """Computes the cartesian product of artifacts."""
+
+    __test__ = False
+
+    def __init__(self,
+                 artifact_sets,
+                 context=TestContext(),
+                 cache_dir=ARTIFACTS_LOCAL_ROOT):
+        """Initialize the cache directory."""
+        self._cache_dir = cache_dir
+        self._context = context
+        # Stores the artifact sets array.
+        self._sets = artifact_sets
+        # ArtifactSet stack pointer.
+        self._set_index = 0
+        if not os.path.exists(cache_dir):
+            os.mkdir(cache_dir)
+
+    @property
+    def sets(self):
+        """Return the artifact sets."""
+        return self._sets
+
+    def download_artifacts(self):
+        """Download all configured artifacts."""
+        for artifact_set in self._sets:
+            for artifact in artifact_set.artifacts:
+                artifact.download(self._cache_dir)
+
+    def _backtrack(self, test_fn, cartesian_product):
+        if len(self._sets) < self._set_index:
+            return
+
+        # Validate solution: tuple element count is equal to
+        # set stack size.
+        if len(self._sets) == len(cartesian_product):
+            self._run_test_fn(cartesian_product, test_fn)
+            return
+
+        current_set = self._sets[self._set_index]
+        for _artifact in current_set.artifacts:
+            # Prepare for recursive call.
+            # Push the current Artifact in the solution.
+            cartesian_product.append(_artifact)
+            # Go 1 level up the ArtifactSet stack.
+            self._set_index += 1
+
+            self._backtrack(test_fn, cartesian_product)
+
+            # Pop the previous artifact from solution.
+            cartesian_product.pop()
+            # Walk down 1 level.
+            self._set_index -= 1
+
+    def _run_test_fn(self, artifacts, test_fn):
+        """Patch context and call test_fn."""
+        for artifact in artifacts:
+            self._context.set_any(artifact.type.value, artifact)
```

Comment:
nit: it would be nice to have a function in `Artifact` class that returns `artifact.type.value`.

Reply:
The type is currently an accessor for the type (which is an enum). Value is an accesor for the Enum object.

Reply:
Yep, I preferred to kind of hide at this level the `type` type (enum), and just call a function from `Artifact` class: `artifact.type_value()`
```
def type_value(self):
        """Return the artifact type value."""
        return self._type.value
```
(but that's pretty subjective I think :-?)

Reply:
We can just use strings instead of enums if we make that change. Artifact types are pretty much opaque to the tests as they are fetched and manipulated through proper APIs.